### PR TITLE
Add ICMP ping exporter

### PR DESCRIPTION
**Pull request by glrf**
## Summary

Adds an ICMP ping exporter that opens an unpriviledged ICMP socket for each configured target. ICMP session ids are handled by the kernel.

Service discovery is handled by periodically reloading DNS. By setting a headless service as a target, this allows kubernetes agnostic service discovery.

The ping codebase is split into two parts:

* Pingers that manage a single connection to a target and consists of a sending go routine and a receiving go routine that communicate through a sync map.

* A manager that resolves DNS entries and starts/stops pingers periodically.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

**Comment by glrf** [comment]
The code is not as well tested as I'd like, especially for error cases, but I'd like to get this running for now and might be able to come back to it later to improve test coverage

File: ping/pinger.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+// NewPinger returns a new instance of pinger
+func NewPinger(ip *net.IPAddr, obs observerVec, inteval, timeout time.Duration) (Pinger, error) {
+	p := pinger{
+		obs: obs,
+		listen: func(network, address string) (net.PacketConn, error) {
+			return icmp.ListenPacket(network, address)
+		},
+	}
+
+	p.sndTicker = time.NewTicker(inteval)
```
Comment (bastjan) [inline ping/pinger.go]: Tickers must be stopped with `.Stop()` before they get garbage collected. I've run into steadily increasing CPU usage before because i did not release tickers.

File: ping/manager.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	if err != nil {
+		return err
+	}
+	pinger, err := m.newPinger(ip, hist, m.pingInterval, m.pingTimout)
+	if err != nil {
+		return err
+	}
+	ctx = logr.NewContext(ctx, l)
+	ctx, cancel := context.WithCancel(ctx)
+	m.pingCtxs[ip.String()] = Context{
```
Comment (bastjan) [inline ping/manager.go:203]: There's now also also the comparable [`net/netip`](https://go.dev/doc/go1.18#netip). But I'm completely fine as is.

File: ping/pinger.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+type observerVec interface {
+	WithLabelValues(...string) prometheus.Observer
+}
+
+type listenPacket func(network, address string) (net.PacketConn, error)
+
+// NewPinger returns a new instance of pinger
+func NewPinger(ip *net.IPAddr, obs observerVec, inteval, timeout time.Duration) (Pinger, error) {
```
Comment (bastjan) [inline ping/pinger.go]:
```suggestion
func NewPinger(ip *net.IPAddr, obs observerVec, interval, timeout time.Duration) (Pinger, error) {
```

File: ping/pinger.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	msgB, err := msg.Marshal(nil)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal ICMP packet: %w", err)
+	}
+
+	if _, err := p.conn.WriteTo(msgB, &p.addr); err != nil {
+		return fmt.Errorf("Failed to send ICMP packet: %w", err)
+	}
+	p.inflight.Store(p.seq, sent)
+	p.seq = p.seq + 1
```
Comment (bastjan) [inline ping/pinger.go:171]: The sequence number is `uint16` according to the [RFC](https://www.rfc-editor.org/rfc/rfc792) so i'm not sure what happens in Go if the int goes over the `uint16` range. I would feel safer if this would be defined as `uint16`.

Also from past experiences the number should be a global counter. 16bit random is not enough, collisions happen frequently and can lead to weird and hard to debug metrics.

**Review by bastjan — CHANGES_REQUESTED** [review]
Mostly LGTM


Both issues from actual past experiences:
- The leaking tickers should be addressed for the pingers.
- The sequence number should be a global counter.

File: main.go [hunk 5]
```diff
@@ -17,16 +24,61 @@ var (
...
 	l.Info("Starting canary..")
+	ctx = logr.NewContext(ctx, l)
+
+	m := ping.NewManager(rttHist, ping.ManagerConfig{
+		Source:         *source,
+		DNSTargets:     *dnsTargets,
+		IPTargets:      *ipTargets,
+		UpdateInterval: *update,
+		PingInterval:   *interval,
+		PingTimout:     *timeout,
```
Comment (bastjan) [inline main.go]:
```suggestion
		PingTimeout:     *timeout,
```

I also don't like Tim 🤗 There are a few other places it is written like this

File: ping/pinger.go (hunk 1 above)
Comment (glrf) [reply ping/pinger.go]: You're right. I totally forgot about that.

File: ping/pinger.go (hunk 4 above)
Comment (glrf) [reply ping/pinger.go:171]: The overflow is a valid question, I suspect that should roll over fine, but you're right we should be explicit.

However I don't really see the reason for the global counter:

* We don't really depend on uniqueness. I really don't yet see the error case. The id is unique per socket that really shouldn't lead to problems
* Also 16bit randomness should be enough. There are only ever max 5 packets in flight per node. The probability of a collision should be low (I'd need to properly calculate that tough)

Comment (glrf) [reply ping/pinger.go:171]: Ah no you're right it might not actually be that unlikely. Not very likely but still.  I guess we can use a global count. I just still don't see the advantage (and it will make testing a bit more annoying..)

Comment (glrf) [reply ping/pinger.go:171]: Oke I finally understood the issue, I think (https://github.com/prometheus/blackbox_exporter/issues/411 has a good example).

However I don't see how a global counter solves this. If we would ping a NATed address through a gateway that also NATs ICMP packets wrong, having a global counter per pod doesn't help us really. The problem would be if two canary pods choose the same seq and id, which is possible.

We would need a global counter for all canary instances, and I  don't really want to go down that rabbithole..

File: ping/manager.go (hunk 2 above)
Comment (glrf) [reply ping/manager.go:203]: That's actually a good input, I forgot that this existed and really aren't a fan of the current solution. However to do this properly I'd need to change quite a bit. I might do this cleanup after the initial release.

File: ping/pinger.go (hunk 4 above)
Comment (bastjan) [reply ping/pinger.go:171]: It really depends on how often targets are added. Every time targets is added there's a chance of overlapping offset. With not much target churn it will be mostly fine. Still, as we are doing stuff that might alert at night I'm on the cautious side here.

The global counter could be just like `time.Now()` and mocked the same.

> We would need a global counter for all canary instances, and I don't really want to go down that rabbithole..

Nope we do not want that 😅 I indeed had the issue with Palo Alto FWs but yeah irrelevant for us.

Comment (bastjan) [reply ping/pinger.go:171]: There is one conn per pinger and the kernel manages the multiplexing through the `ID` field. Thus the session ID is not very relevant for us. Nothing will break with overlapping session ids.

Comment (glrf) [reply ping/pinger.go:171]: I found the kernel patch for this: https://lwn.net/Articles/420800/

TL;DR the way we currently use this, the kernel chooses a free ICMP ID similarly to choosing a free UDP port when opening a socket with port 0. So each `conn` has a unique ID and the kernel will multiplex them. (We could also use custom IDs by setting the port in the UDP address I think, but I don't see a use-case for that right now)

**Review by bastjan — APPROVED** [review]
LGTM 🚀

File: ping/pinger.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		if ctx.Err() != nil {
+			return nil
+		}
+
+		if _, ok := p.inflight.LoadAndDelete(uint16(res.Seq)); ok {
+			received := now()
+			sent := time.UnixMicro(int64(binary.BigEndian.Uint64(res.Data)))
+			l.V(1).Info("got reflection", "id", res.ID, "seq", res.Seq, "sent", sent, "received", received, "latency", received.Sub(sent))
+			p.obs.WithLabelValues("completed").Observe(received.Sub(sent).Seconds())
+			p.inflight.Delete(res.Seq)
```
Comment (bastjan) [inline ping/pinger.go]: This is not needed right? 😅

Comment (glrf) [reply ping/pinger.go]: Argh, yes some refactoring troubles :see_no_evil:
